presto-main/src/main/java/com/facebook/presto/metadata/FunctionExtractor.java

```diff
@@ -64,6 +65,11 @@ public static List<? extends SqlFunction> extractFunctions(Class<?> clazz)
             return SqlInvokedScalarFromAnnotationsParser.parseFunctionDefinition(clazz);
         }
 
+        List<? extends SqlFunction> annotatedFunctionInstances = parseFunctionsFromAnnotation(clazz);
```

Comment:
Can you explain why we cannot just use SqlAggregationFunction::createFunctionByAnnotations ?

Reply:
If you implement an aggregation function with annotations through `@AggregationFunction` the type constraints _must_ be specified in your subsequent aggregation function annotations. Thus, if you wanted to implement a function for more than one type, you need a method and annotation for each specific input type. As an example of multiple accepted types using the annotations, see something like [`AverageAggregations`](https://github.com/prestodb/presto/blob/0.284/presto-main/src/main/java/com/facebook/presto/operator/aggregation/AverageAggregations.java)

However, if you want to implement a function which can be specialized for _any_ type, then you need to extend the `SqlAggregationFunction` class and then pass an _instance_ of that function to the `BuiltInFunctionAndTypeManager#function` method. An example of this type of implementation can be found in [`ChecksumAggregationFunction`](https://github.com/prestodb/presto/blob/0.284/presto-main/src/main/java/com/facebook/presto/operator/aggregation/ChecksumAggregationFunction.java), but there are also many more implemented this way. 

However, since you need an instance of the aggregation class to register a generic aggregation function, the current plugin interface is insufficient because all you can do is pass a `Class<?>`. Thus, I added the ability to extract a static instance of the function which the `FunctionAndTypeManager` can use execute the function on any generic type.

Reply:
I'm not sure I follow--can you reference `BuildSetDigestAggregation` and see if this does what you need it to do?  It is not specialized for any particular type.

Reply:
Thanks for the pointer Tim - I didn't come across this before. . Most of the other aggregation function that accept a type parameter that I found previously all seemed to be implemented without the `@TypeParameter` annotation. I'll try to implement it using that

Reply:
I made an intermediate PR which will allow us to simplify this implementation in #21316

Reply:
The implementation has been updated to use the annotation framework and the rest of the changes around function parsing were reverted